### PR TITLE
Refactor Dynamic Node Callbacks

### PR DESCRIPTION
**Pull request by camielvs**
## Description

`generateDynamicNodeCallbacks` has now been replaced with `convertNodeCallbacksToTaskCallbacks`, which is much simpler in logic and has clearer typing.

This relates to the automatic generation of Task Callbacks during node creation, and, to make the distinction clearer, the input `NodeData` type has been separated from the internal `TaskNodeData` type.

**Background explanation**
A Task Node's callbacks are defined at the Flow-Canvas level in `useNodeCallbacks` and require the `taskId` and `nodeId` to be passed in so ReactFlow & ComponentSpec can be updated accordingly. However, within a task node, where these callbacks are executed, `taskId` and `nodeId` are already known and are static. Thus, we currently have a transform that occurs when Task Nodes are created to inject the known ids into the callback function. This saves us from having to manually pass in the ids when executing the callbacks within a node (e.g. in TaskNodeProvider).

This PR simplifies that conversion function significantly and institutes a clear division between the internal TaskNode data and the external node data needed to create the task node.

This PR is intended to start laying the groundwork for a node compatibility layer to be introduced, which will ultimately involve id generation at the task node creation level. Having simplified task creation code will make this process easier.

No change to app functionality.

Note/Follow-up:
- Previously `generateDynamicNodeCallbacks` did the conversion dynamically. this new system requires the methods to be manually coded. Consider if there is a way to go back to a dynamic model so that any new methods don't need to be manually added

## Related Issue and Pull requests

Closes https://github.com/Shopify/oasis-frontend/issues/334

## Type of Change

- [x] Cleanup/Refactor

## Checklist

- [ ] I have tested this does not break current pipelines / runs functionality
- [ ] I have tested the changes on staging

## Test Instructions

1. Create Task Nodes
- Open a pipeline editor
- Add multiple task nodes to the canvas
- Verify nodes render correctly with proper positioning

2. Test Node Callbacks
- setArguments: Modify task arguments and verify they persist
- setAnnotations: Move/resize nodes and verify position updates save
- onDelete: Delete a task node and verify it's removed from the spec
- onDuplicate: Duplicate a node (with/without selection) and verify the copy is created
- onUpgrade: If applicable, upgrade a component and verify the node updates

3. Create Input/Output Nodes
- Create/verify input nodes render correctly
- Create/verify output nodes render correctly
- Connect task nodes to input/output nodes

4. Edge Cases
- Load a complex existing pipeline to verify node reconstruction works
- Test with nodes that have no callbacks defined
- Test with read-only nodes

5. Regression Check
- Run through a complete pipeline creation workflow
- Save, reload, and verify state persistence
- Execute a pipeline and verify it runs successfully

Expected Results
All functionality should behave identically to before the refactor. No visual or functional changes should be observed.

### CODE REVIEW

File: src/utils/nodes/createTaskNode.ts [hunk 1]
```diff
@@ -1,34 +1,37 @@
...
+  const { nodeCallbacks, ...data } = nodeData;
 
   const position = extractPositionFromAnnotations(taskSpec.annotations);
   const nodeId = taskIdToNodeId(taskId);
 
   // Inject the taskId and nodeId into the callbacks
-  const nodeCallbacks = nodeData.nodeCallbacks;
-  const dynamicCallbacks = generateDynamicNodeCallbacks(nodeId, nodeCallbacks);
+  const ids = { taskId, nodeId };
+  const taskCallbacks = convertNodeCallbacksToTaskCallbacks(ids, nodeCallbacks);
```
Comment (Mbeaulne) [inline src/utils/nodes/createTaskNode.ts]: NIT: We aren't using this `ids` variable anywhere, so It think generating the object directly in the `convertNodeCallbacksToTaskCallbacks` might read better. 

``` const taskCallbacks = convertNodeCallbacksToTaskCallbacks({ taskId, nodeId }, nodeCallbacks);```

File: src/utils/nodes/taskCallbackUtils.ts [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+    setAnnotations: (annotations) =>
+      nodeCallbacks.setAnnotations?.(ids, annotations),
+    onDelete: () => nodeCallbacks.onDelete?.(ids),
+    onDuplicate: (selected) => nodeCallbacks.onDuplicate?.(ids, selected),
+    onUpgrade: (newComponentRef) =>
+      nodeCallbacks.onUpgrade?.(ids, newComponentRef),
+  };
+};
+
+// Empty callbacks
```
Comment (Mbeaulne) [inline src/utils/nodes/taskCallbackUtils.ts]: I think we can remove this comment

**Review by Mbeaulne — APPROVED** [review]

File: src/utils/nodes/taskCallbackUtils.ts [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+    setAnnotations: (annotations) =>
+      nodeCallbacks.setAnnotations?.(ids, annotations),
+    onDelete: () => nodeCallbacks.onDelete?.(ids),
+    onDuplicate: (selected) => nodeCallbacks.onDuplicate?.(ids, selected),
+    onUpgrade: (newComponentRef) =>
+      nodeCallbacks.onUpgrade?.(ids, newComponentRef),
+  };
+};
+
+export const createEmptyTaskCallbacks = (): TaskCallbacks => ({
```
Comment (maxy-shpfy) [inline src/utils/nodes/taskCallbackUtils.ts]: NIT: This seems to be internal (flagged by knip), so unless we use it externally, `export` may be removed

File: src/types/taskNode.ts [hunk 4]
```diff
@@ -6,26 +6,31 @@ import type {
 
 import type { Annotations } from "./annotations";
 
+export type TaskType = "task" | "input" | "output";
+
+export interface NodeData extends Record<string, unknown> {
+  readOnly?: boolean;
+  connectable?: boolean;
```
Comment (maxy-shpfy) [inline src/types/taskNode.ts:13]: For future: do we have any instance of Graph to be readonly, BUT connectable? I feel that this is single property "readOnly".

File: src/types/taskNode.ts [hunk 5]
```diff
@@ -6,26 +6,31 @@ import type {
 
 import type { Annotations } from "./annotations";
 
+export type TaskType = "task" | "input" | "output";
+
+export interface NodeData extends Record<string, unknown> {
```
Comment (maxy-shpfy) [inline src/types/taskNode.ts:11]: For future: Do we need to have more arbitrary properties on this object? extending it from `Record` makes it harder to type causing weird code to appear down the road, e.g.

TaskNode: 

```typescript
const typedData = useMemo(() => data as TaskNodeData, [data]);
```

File: src/types/taskNode.ts [hunk 6]
```diff
@@ -6,26 +6,31 @@ import type {
...
+  nodeCallbacks?: NodeCallbacks;
+}
+
 export interface TaskNodeData extends Record<string, unknown> {
   taskSpec?: TaskSpec;
   taskId?: string;
   readOnly?: boolean;
   isGhost?: boolean;
   connectable?: boolean;
   highlighted?: boolean;
```
Comment (maxy-shpfy) [inline src/types/taskNode.ts:23]: For future: `highlighted` might not belong to the Node properties/data, but I'm not sure yet.

**Review by maxy-shpfy — APPROVED** [review]
Tested, no immediate issues found

File: src/types/taskNode.ts (hunk 4 above)
Comment (camielvs) [reply src/types/taskNode.ts:13]: It's more the opposite scenario that's the concern: the graph is not readonly and node connection is disabled - specifically this lock button



I've been growing increasingly discontent with "readOnly" and how it's used specifically for the `run` route. I think in future we will want to take a look and properly make `readOnly` an actual read-only state regardless of the route.

File: src/types/taskNode.ts (hunk 6 above)
Comment (camielvs) [reply src/types/taskNode.ts:23]: Most likely the node overlay or highlighter you made would solve this but we haven't gotten around to looking into it.

File: src/types/taskNode.ts (hunk 5 above)
Comment (camielvs) [reply src/types/taskNode.ts:11]: I vaguely recall trying to not extend and ReactFlow threw up a huge amount of type errors. There may be some larger changes need to our type structure to make this happen, but I do agree with the idea.

File: src/utils/nodes/taskCallbackUtils.ts (hunk 3 above)
Comment (camielvs) [reply src/utils/nodes/taskCallbackUtils.ts]: That's true. It's used immediately in the next PR and it made more sense to add it here with its counterpart. 99.99999% chance both PRs will be merged together so I don't think it's worth updating or moving the code.

**Comment by camielvs** [comment]
Closed in favour of v2 Editor migration. See https://github.com/TangleML/tangle-ui/pull/2028
